### PR TITLE
Make exit codes configurable.

### LINKED ISSUE #790: CodeClimate engine - exit code
Currently, Reek exits with a status of 2 when it detect smells.

When running Reek as a Code Climate engine (see #763), the engine needs to exit with a code of 0 when there are smells.

I'm trying to think how this can be elegantly handled in Reek.

The most direct way would be to add a conditional in `Application#report_smells` which reads the `options.report_format`, but that's essentially a type check smell.

I was thinking a better approach might be to set the status codes on a per-formatter basis, so each report would have methods for `#success_exit_code`, `#error_exit_code` and `#smells_exit_code`.

This has the advantage that any future engines can control the exit codes, but it's perhaps a premature optimisation since this is currently only needed for Code Climate.

Any thoughts?

### PR DESCRIPTION
**Pull request by troessner**
Fixes #790

### CODE REVIEW

File: features/step_definitions/sample_file_steps.rb [hunk 1]
```diff
@@ -1,3 +1,22 @@
...
+
+Given(/^the clean file 'clean.rb'$/) do
+  contents = <<-EOS.strip_heredoc
+    # Explanatory comment
+    class Clean
+      def foo; end
+    end
+  EOS
+  write_file('clean.rb', contents)
+end
```
Comment (troessner) [inline features/step_definitions/sample_file_steps.rb]: I believe we could and should reduce and unify all of our used code samples since they are quite a few that overlap. But that would be a separate pull request.

File: lib/reek/cli/application.rb [hunk 2]
```diff
@@ -11,53 +11,45 @@ module CLI
...
       def report_smells
-        self.status = STATUS_SMELLS
+        self.status = options.failure_exit_code
       end
 
       private
 
-      def error_occured?
-        status == STATUS_ERROR
-      end
```
Comment (troessner) [inline lib/reek/cli/application.rb]: That method could never be called at runtime.

File: lib/reek/cli/application.rb [hunk 3]
```diff
@@ -11,53 +11,45 @@ module CLI
...
-      STATUS_SMELLS  = 2
       attr_reader :configuration
 
       private_attr_accessor :status
       private_attr_reader :command, :options
 
       def initialize(argv)
-        @status = STATUS_SUCCESS
         @options = configure_options(argv)
+        @status = options.success_exit_code
```
Comment (troessner) [inline lib/reek/cli/application.rb]: `status` now needs to be initialized after `options`, hence the moving of the lines.

File: lib/reek/cli/options.rb [hunk 4]
```diff
@@ -9,9 +9,14 @@ module CLI
     #
     # See {file:docs/Command-Line-Options.md} for details.
     #
-    # :reek:TooManyInstanceVariables: { max_instance_variables: 7 }
+    # :reek:TooManyInstanceVariables: { max_instance_variables: 9 }
     # :reek:Attribute: { enabled: false }
+    #
     class Options
+      DEFAULT_SUCCESS_EXIT_CODE  = 0
+      DEFAULT_ERROR_EXIT_CODE    = 1
```
Comment (troessner) [inline lib/reek/cli/options.rb]: DEFAULT_ERROR_EXIT_CODE is **not** configurable at the moment. I just put it here for consistency reasons. 
We exit the hard way using this code only at 2 places:
1.) https://github.com/troessner/reek/pull/808/files#diff-8cb50151ad3a458ecb02fe222bb330eaR52
2.) https://github.com/troessner/reek/pull/808/files#diff-8cb50151ad3a458ecb02fe222bb330eaR45

and that's where we're actually parsing options. So making them configurable right now would be completely pointless.
If we start to introduce this error code across then we should make it configurable as well.
And we should probably. Thinking about #809

File: lib/reek/cli/options.rb [hunk 5]
```diff
@@ -19,16 +24,20 @@ class Options
...
-        @show_links       = true
-        @smells_to_detect = []
-        @colored          = color_support?
+        @argv              = argv
+        @parser            = OptionParser.new
+        @report_format     = :text
+        @location_format   = :numbers
+        @show_links        = true
+        @smells_to_detect  = []
+        @colored           = color_support?
```
Comment (troessner) [inline lib/reek/cli/options.rb]: Changes here and above are just indentation changes.

**Comment by chastell** [comment]
LGTM, but @mvz didn’t comment on this approach in #790, so I’ll defer merging to him.

**Comment by mvz** [comment]
I'm not sure all this is needed, but technically I have no problems with this approach. Merging.

**Comment by troessner** [comment]
What do you mean with "all this is needed"? It's a prerequisite for the codeclimate integration.

**Comment by mvz** [comment]
My thoughts were:
* We could have used just one hard-code exit code for the code climate reporter (less flexible, but enough for that use case).
* We may not need to customize the success exit code.
On the plus side:
* This is a flexible and consistent solution that will allow people to customize codes to their heart's content.

**Comment by troessner** [comment]
>> This is a flexible and consistent solution that will allow people to customize codes to their heart's content.

:+1: ;)
